doc/manual/contract.md

```diff
@@ -0,0 +1,329 @@
+# Contract Programming
+
+Nit supports [contract programming](https://en.wikipedia.org/wiki/Design_by_contract). Contracts can be seen as the materialization of specifications in the code itself. The idea is to define the behavior of a property using conditions which will be evaluated on runtime. A contract define the relations between a property and all the potential callers. If one of the two parties (callee/caller) does not respect the defined rules, the contract is considered to be broken and will cause a runtime error.
+
+You can put contract on different elements of the language: methods, attributes, interfaces and classes in order to verify the conformity of the implementation compared with the specification. Different type of contracts exists; the preconditions (`expect`) which check the conditions before execution, the postconditions (`ensure`) which checks the conditions after execution and the invariants (`invariant`) which checks that the conditions remain unchanged.
+
+The implementation of contracts is recommended when you are dealing with business rules. For example when you are developing a library it may be necessary to specify contracts to guarantee valid use of the API. Also, contracts support inheritance, which guarantees that contracts will be satisfied in future specification.
+
+Let us take a look at the `Stack` class bellow:
+
+~~~
+	class Stack[T]
+		invariant(size >= 0)
+
+		var data = new Array[T]
+
+		var size = 0
+
+		fun is_empty: Bool do return size == 0
```

Comment:
```
ensure(result == (size == 0))
```

I think wirh design by contracts you're supposed to give post conditions of your derived requests based on on the primitive ones.